inc/marketplace/api/plugins.class.php

```diff
@@ -0,0 +1,259 @@
+               'tags',
+               'tag',
+               'authors',
+               'author',
+               'version',
+               'message',
+            ])
+         ]
+      ], 'POST');
+
+      $token = json_decode($response->getBody(), true);
+      $this->accessToken = $token['access_token'];
+      $GLPI_CACHE->set('marketplace_accessToken', $this->accessToken, (int)$token['expires_in']);
+
+      return true;
+   }
+
+   function getPlugins(bool $force = false) {
+      global $GLPI_CACHE;
+
+      if (!$force && $GLPI_CACHE->has('marketplace_all_plugins')) {
+         return $GLPI_CACHE->get('marketplace_all_plugins');
+      }
+
+      $plugins = [];
+      $step    = 50;
+      $i       = 0;
+      do {
+         $response = $this->request('plugin', [
+            'headers' => [
+               'X-Range' => ($i * $step)."-".(($i + 1) * $step - 1),
+            ],
+         ]);
+
+         if ($current_plugins = json_decode($response->getBody(), true)) {
+            $plugins = array_merge($plugins, $current_plugins);
+         }
+
+         $i++;
+      } while ($current_plugins !== false && count($current_plugins));
+
+      // replace keys indexes by system names
+      $plugins_keys = array_column($plugins, 'key');
+      $plugins      = array_combine($plugins_keys, $plugins);
+
+      $GLPI_CACHE->set('marketplace_all_plugins', $plugins, HOUR_TIMESTAMP);
+      return $plugins;
+   }
+
+   function getValidPlugins(bool $force = false) {
+      global $GLPI_CACHE;
+
+      if (!$force && $GLPI_CACHE->has('marketplace_valid_plugins')) {
+         return $GLPI_CACHE->get('marketplace_valid_plugins');
+      }
+
+      $plugins = $this->getPlugins($force);
+      $list_url = [];
+      $list_versions = [];
+      $valid_plugins = array_filter($plugins, function($plugin) use(&$list_url, &$list_versions) {
+         $url_exists = false;
+         foreach ($plugin['versions'] as $version) {
+            if (isset($version['download_url'])
+                && strlen($version['download_url']) > 0
+                && $this->matchGlpiVersion($version['compatibility'])) {
+               $url_exists = true;
+               $list_url[$plugin['key']] = $version['download_url'];
+               $list_versions[$plugin['key']] = $version['num'];
+               break;
+            }
+         }
+
+         return $url_exists;
+      });
+
+      // push installation url in base key
+      foreach ($valid_plugins as &$plugin) {
+         $plugin['installation_url'] = $list_url[$plugin['key']] ?? "";
+         $plugin['version'] = $list_versions[$plugin['key']] ?? "";
+      }
+
+      $GLPI_CACHE->set('marketplace_valid_plugins', $valid_plugins, HOUR_TIMESTAMP);
+      return $valid_plugins;
+   }
+
+
+   function matchGlpiVersion(string $constraint = ""):bool {
+      $glpi_version  = defined('GLPI_PREVER') ? GLPI_PREVER : GLPI_VERSION;
+      $glpi_version = "9.4";
+
+      return Semver::satisfies($glpi_version, $constraint);
```

Comment:
No.
GLPI doesn't follow SemVer (e.g. method removed in 9.4 so not compatible with 9.3)

Reply:
Indeed.

The idea here is not to tells that GLPI repect Semver, but to use an existing tool to check compatibility declared in plugins API against GLPI version.
So if a plugins tells it is compatible with `~9.3.2 | ~9.4.0`, we will be able to validate this constraint even if GLPI does not follow Semver.

Reply:
but `^9.4` won't work

Reply:
Indeed, but we expect plugin developpers to know that GLPI does not follow Semver and use strong constraints to handle with that.
Other projects does not follow Semver but can still be managed using its notation. For example, Laravel prior to version 6 was not following semver, but it was still possible to fetch its components using a composer.json configuration file.